pkg/cli/admin/release/codesign/machoresign.go

```diff
@@ -0,0 +1,191 @@
+package codesign
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"debug/macho"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+func ResignMacho(path string, asArchive bool, command string, links []string) error {
+	// Update Mach-O code signature, if any.
+	// After injecting version or image, binaries become invalidated on mac os arm64
+	// architectures and we have to sign them according to the new content.
+	// Signing is handled exactly the similar way how go handles same situation.
+	if !asArchive {
+		return signSingleFile(path)
+	}
+
+	// archived files are not seekable and go re-sign process needs writerat and readerat
+	// interfaces. Therefore, we are extracting these files into temp location, signing there,
+	// overriding current path with this new archive file.
+	return signTarFile(path, command, links)
+}
+
+func signSingleFile(path string) error {
+	file, err := os.OpenFile(path, os.O_RDWR, 0755)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	wr := io.WriterAt(file)
+	if machoFile, cmd, ok := findMachoCodeSignature(wr); ok {
+		if Size(int64(cmd.Dataoff), "a.out") == int64(cmd.Datasize) {
+			// Update the signature if the size matches, so we don't need to
+			// fix up headers. Binaries generated by the Go linker should have
+			// the expected size. Otherwise skip.
+			text := machoFile.Segment("__TEXT")
+			cs := make([]byte, cmd.Datasize)
+			Sign(cs, wr.(io.Reader), "a.out", int64(cmd.Dataoff), int64(text.Offset), int64(text.Filesz), machoFile.Type == macho.TypeExec)
+			if _, err := wr.WriteAt(cs, int64(cmd.Dataoff)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func signTarFile(path, command string, links []string) error {
```

Comment:
> Does the newly written tar file respects that same file permissions of the original tar? 

Yes, it is one of the important thing we need to consider and this PR preserves file permissions of original tar.

> Can we always expect the release image will always provide tar and not e.g. gzip archives?

That is a good point. In my initial draft, I had also implemented for `zip` archives. But after digging through the code, I noticed that `zip` is only supported for Windows(see [here](https://github.com/openshift/oc/blob/a1119e4cd75df284e74bfd731f22b65b2dac0c75/pkg/cli/admin/release/extract_tools.go#L228)) and we are only interested in macos arm64.

Reply:
My understanding is the `oc adm release extract` command is executed on a linux machine/container. So zip, gzip, tar and other formats are supported. Not necessarily present.

Reply:
Supported targets are listed in here https://github.com/openshift/oc/blob/a1119e4cd75df284e74bfd731f22b65b2dac0c75/pkg/cli/admin/release/extract_tools.go#L173. In reality, we support `zip` for Windows and `tar.gz` for the rest.

IMO, since we are re-signing archive files we produce as `tar.gz`, we can re-sign only `tar.gz` and ignore other formats.